File: SharedLibrary/Shared/Objects/MActionResult.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using JetBrains.Annotations;
using SharedLibrary.Enums;
using SharedLibrary.Interfaces;

namespace SharedLibrary
{
	[Android.Runtime.Preserve(AllMembers = true)]
	public class MActionResult<T> where T : class
	{
		[CanBeNull]
		private T _content;

		public T Content
		{
			get => _content;
			set => this._content = value;
		}

		public Exception Exception { get; set; }
		public bool IsSuccess => (Status == StatusCode.OK && Content != null);
		public StatusCode Status { get; set; }
		public Version Version { get; set; } = System.Version.Parse("1.0.0");
		public DateTime Date { get; set; }

		public MActionResult() { }

		public MActionResult(StatusCode status) : this(status, default(T), null) { }

		public MActionResult(StatusCode status, T result) : this(status, result, null) { }

		public MActionResult(StatusCode status, Exception exception) : this(status, default(T), exception)
		{
			if (exception == null) {
				throw new ArgumentNullException("Exception is null");
			}
		}

		public MActionResult(StatusCode status, T result, Exception exception)
		{
			if (result == null) {
				this.Content = default(T);
			} else {
				Type currentType = typeof(T);

				bool isCollection = currentType.IsGenericType &&
									currentType.GetGenericTypeDefinition() == typeof(List<>) &&
									result.GetType().GetGenericArguments().Single().GetInterfaces()
										  .Contains(typeof(IDbObject));

				if (!(currentType.GetInterfaces().Contains(typeof(IDbObject)) || isCollection)) {
					throw new ArgumentException("content",
												"Type must Implement IDbObject be List<>, but was " +
												currentType.FullName);
				}

				this.Content = result;
			}

			this.Status = status;
			this.Exception = exception;
			this.Date = DateTime.Now;
		}

		public string GetStatusMessage()
		{
			switch (this.Status) {
				case StatusCode.OK:
					return "Vše v pořádku";
				case StatusCode.SeeException:
					string s = "Vyskytla se chyba";
					if (this.Exception != null) {
						s += $", podrobnosti: {this.Exception.Message}";
					}
					return s;
				case StatusCode.NoPassword:
					return "Nemáte vytvořené heslo. Použijte funkci zapomenuté heslo";
				case StatusCode.Forbidden:
					return "Zakázáno";
				case StatusCode.NotFound:
					return "Nenalezeno";
				case StatusCode.Timeout:
					return "Vypršel časový limit na danou operaci";
				case StatusCode.InternalError:
					return "Vyskytla se chyba na straně serveru";
				case StatusCode.NotValidID:
					return "Poskytnuté ID není platné";
				case StatusCode.AlreadyExists:
					return "Tento záznam již existuje";
				case StatusCode.InvalidInput:
					return "Neplatné vstupní parametry";
				case StatusCode.NotEnabled:
					return "Uživatel není aktivovaný";
				case StatusCode.WrongPassword:
					return "Neplatné jméno nebo  heslo";
				case StatusCode.ExpiredPassword:
					return "Zadané heslo již není platné";
				case StatusCode.InsufficientPermissions:
					return "Na tutu akci nemáte dostatečné oprávnění";
				case StatusCode.CannotTransfer:
					return "Vybraného žolíka nelze přenést";
				case StatusCode.CannotSplit:
					return "Vybraného žolík nelze rozdělit";
				case StatusCode.JustALittleError:
					return "Vyskytla se chyba, ale pouze malá :)";
				default:
					throw new ArgumentOutOfRangeException();
			}
		}
	}
}

File: MobileApp/zoliky_teachers/lib/utils/api/models/universal/MActionResult.dart
import 'package:zoliky_teachers/utils/api/Extends.dart';
import 'package:zoliky_teachers/utils/api/enums/StatusCode.dart';

class MActionResult<T> {
  T content;
  Object exception;
  bool get isSuccess =>
      (status.value == StatusCode.OK.value && content != null);
  StatusCode status;
  DateTime date;

  MActionResult();

  MActionResult<T> ctorOnlyStatus(StatusCode status) {
    return ctorContentAndException(status, null, null);
  }

  MActionResult<T> ctorWithContent(StatusCode status, T result) {
    return ctorContentAndException(status, result, null);
  }

  MActionResult<T> ctorWithexception(Object ex) {
    return ctorContentAndException(StatusCode.SeeException, null, ex);
  }

  MActionResult<T> ctorContentAndException(
      StatusCode status, T result, Object ex) {
    this.status = status;
    this.exception = ex;
    this.content = result;
    return this;
  }

  String getStatusMessage() {
    if (this.status == StatusCode.OK) {
      return "Vše v pořádku";
    } else if (this.status == StatusCode.SeeException) {
      String s = "Vyskytla se chyba";
      if (this.exception != null) {
        s += ", podrobnosti){ ${this.exception.toString()}";
      }
      return s;
    } else if (this.status == StatusCode.NoPassword) {
      return "Nemáte vytvořené heslo. Použijte funkci zapomenuté heslo";
    } else if (this.status == StatusCode.Banned) {
      return "Váš účet je zabanován";
    } else if (this.status == StatusCode.Forbidden) {
      return "Zakázáno";
    } else if (this.status == StatusCode.NotFound) {
      return "Nenalezeno";
    } else if (this.status == StatusCode.Timeout) {
      return "Vypršel časový limit na danou operaci";
    } else if (this.status == StatusCode.InternalError) {
      return "Vyskytla se chyba na straně serveru";
    } else if (this.status == StatusCode.NotValidID) {
      return "Poskytnuté ID není platné";
    } else if (this.status == StatusCode.AlreadyExists) {
      return "Tento záznam již existuje";
    } else if (this.status == StatusCode.InvalidInput) {
      return "Neplatné vstupní parametry";
    } else if (this.status == StatusCode.NotEnabled) {
      return "Uživatel není aktivovaný";
    } else if (this.status == StatusCode.EmailNotConfirmed) {
      return "Email není potvrzený";
    } else if (this.status == StatusCode.WrongPassword) {
      return "Neplatné jméno nebo  heslo";
    } else if (this.status == StatusCode.ExpiredPassword) {
      return "Zadané heslo již není platné";
    } else if (this.status == StatusCode.InsufficientPermissions) {
      return "Na tutu akci nemáte dostatečné oprávnění";
    } else if (this.status == StatusCode.CannotTransfer) {
      return "Vybraného žolíka nelze přenést";
    } else if (this.status == StatusCode.CannotSplit) {
      return "Vybraného žolík nelze rozdělit";
    } else if (this.status == StatusCode.JustALittleError) {
      return "Vyskytla se chyba, ale pouze malá :)";
    }
    return "Neznámá chyba";
  }

  MActionResult.fromJson(Map<String, dynamic> _json)
      : content = _json["Content"],
        exception = _json["Exception"],
        status = StatusCode(_json["Status"]),
        date = Extends.parseDateTime(_json["Date"].toString());

  MActionResult.fromJsonWthContent(Map<String, dynamic> _json)
      : exception = _json["Exception"],
        status = StatusCode(_json["Status"]),
        date = Extends.parseDateTime(_json["Date"].toString());
}
